File: backend/save/save_rebuilt_3d.php
<?php
header('Content-Type: application/json; charset=utf-8');
try {
    $raw = file_get_contents('php://input');
    if ($raw === false || strlen($raw) === 0) {
        http_response_code(400);
        echo json_encode([ 'ok' => false, 'error' => 'empty body', 'engine' => 'php' ], JSON_UNESCAPED_UNICODE);
        exit;
    }
    $decoded = json_decode($raw, true);
    if ($decoded === null) {
        http_response_code(400);
        echo json_encode([ 'ok' => false, 'error' => 'invalid json', 'engine' => 'php' ], JSON_UNESCAPED_UNICODE);
        exit;
    }
    $baseDir = dirname(__DIR__, 2) . '/海缆数据/重构的数据';
    if (!is_dir($baseDir)) {
        if (!mkdir($baseDir, 0775, true) && !is_dir($baseDir)) {
            http_response_code(500);
            echo json_encode([ 'ok' => false, 'error' => 'mkdir failed', 'engine' => 'php' ], JSON_UNESCAPED_UNICODE);
            exit;
        }
    }
    $items = [];
    if (isset($decoded['items']) && is_array($decoded['items'])) {
        $items = $decoded['items'];
    } elseif (is_array($decoded)) {
        $items = $decoded;
    }
    $savedAt = (int) round(microtime(true) * 1000);
    $payload = [ 'version' => 1, 'savedAt' => $savedAt, 'items' => array_values($items) ];
    $json = json_encode($payload, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
    if ($json === false) {
        throw new RuntimeException('encode failed');
    }
    $path = $baseDir . '/3d海缆数据.json';
    $ok = file_put_contents($path, $json);
    if ($ok === false) {
        throw new RuntimeException('write failed');
    }
    echo json_encode([ 'ok' => true, 'path' => $path, 'engine' => 'php' ], JSON_UNESCAPED_UNICODE);
} catch (Throwable $e) {
    http_response_code(500);
    echo json_encode([ 'ok' => false, 'error' => $e->getMessage(), 'engine' => 'php' ], JSON_UNESCAPED_UNICODE);
}

File: backend/save/save_rebuilt_3d.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function
import sys, json, os, time, io


def respond(payload):
    print(json.dumps(payload, ensure_ascii=False), end='')


def ensure_dir(path):
    if not os.path.isdir(path):
        os.makedirs(path)


def main():
    raw = sys.stdin.read()
    try:
        data = json.loads(raw)
    except Exception:
        respond({'ok': False, 'error': 'invalid json'})
        return
    base = os.path.join(os.path.dirname(os.path.dirname(__file__)), '海缆数据', '重构的数据')
    ensure_dir(base)
    items = []
    if isinstance(data, dict) and isinstance(data.get('items'), list):
        items = data['items']
    elif isinstance(data, list):
        items = data
    saved_at = int(time.time() * 1000)
    payload = {'version': 1, 'savedAt': saved_at, 'items': list(items)}
    path = os.path.join(base, '3d海缆数据.json')
    with io.open(path, 'w', encoding='utf-8') as f:
        json.dump(payload, f, ensure_ascii=False)
    respond({'ok': True, 'path': path})


if __name__ == '__main__':
    main()
